### PR TITLE
Arena setup

### PR DESCRIPTION
**Pull request by phadej**

### CODE REVIEW

**Comment by phadej** [comment]
This is ready. I run micro benchmarks

```plain
Benchmark                                                                                                    main      arena            
Bench.Database.LSMTree.Internal.Lookup/2_000_000 entries, 256 negative lookups/Bloomfilter query             0.779e-5  0.778e-5   -0.09%
Bench.Database.LSMTree.Internal.Lookup/2_000_000 entries, 256 negative lookups/Compact index search          2.940e-7  9.549e-7 +224.73%
Bench.Database.LSMTree.Internal.Lookup/2_000_000 entries, 256 negative lookups/Lookup preparation in memory  0.813e-5  0.876e-5   +7.76%
Bench.Database.LSMTree.Internal.Lookup/2_000_000 entries, 256 negative lookups/Lookups in IO                 2.907e-5  3.258e-5  +12.08%
Bench.Database.LSMTree.Internal.Lookup/2_000_000 entries, 256 negative lookups/Perform intra-page lookups    1.253e-6  1.320e-6   +5.37%
Bench.Database.LSMTree.Internal.Lookup/2_000_000 entries, 256 negative lookups/Submit IOOps                  1.638e-5  2.204e-5  +34.56%
Bench.Database.LSMTree.Internal.Lookup/2_000_000 entries, 256 positive lookups/Bloomfilter query             1.601e-5  1.590e-5   -0.66%
Bench.Database.LSMTree.Internal.Lookup/2_000_000 entries, 256 positive lookups/Compact index search          0.504e-4  1.395e-4 +176.80%
Bench.Database.LSMTree.Internal.Lookup/2_000_000 entries, 256 positive lookups/Lookup preparation in memory  0.673e-4  1.746e-4 +159.55%
Bench.Database.LSMTree.Internal.Lookup/2_000_000 entries, 256 positive lookups/Lookups in IO                 0.540e-3  0.654e-3  +21.08%
Bench.Database.LSMTree.Internal.Lookup/2_000_000 entries, 256 positive lookups/Perform intra-page lookups    0.398e-4  0.458e-4  +14.98%
Bench.Database.LSMTree.Internal.Lookup/2_000_000 entries, 256 positive lookups/Submit IOOps                  0.393e-3  0.432e-3   +9.86%
Geometric mean                                                                                               2.037e-5  2.868e-5  +40.79%
```

It's not unexpected that `indexSearches` "Compact index search" is slower, as a) it is quick function b) the fresh allocation with (not-yet cached) page allocation will have some overhead. It (hopefully) will be faster asymptotically.

TL;DR in micro benchmark this optimisation may look worse. And especially at this stage, where it's only a setup.

File: src/Data/Arena.hs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+-- | Create unmanaged arena
+--
+-- Never use this in non-tests code.
+withUnmanagedArena :: PrimMonad m => (Arena (PrimState m) -> m a) -> m a
+withUnmanagedArena k = do
+    mgr <- newArenaManager
+    withArena mgr k
+
+allocateFromArena :: PrimMonad m => Arena (PrimState m)-> Size -> Alignment -> m (Offset, MutableByteArray (PrimState m))
+allocateFromArena (Arena mvar) !size !alignment = do
```
Comment (dcoutts) [inline src/Data/Arena.hs:71]: So this is quite general, supporting arbitrary size and alignment. (I know that's what an arena allocator can do in general.)

Our use case is simpler, we should just need allocations that are all of the same size (4k) and _not_ need to be able to allocate contiguous spans of pages.

Yes, currently the `indexSearches` currently does a single IOOpRead for a span of pages into a single contigious buffer, but I think we should aim to change that to issue multiple reads for each individual page into separate buffers.

That way the underlying allocator only has to deal with a single size of allocation, and so we can use a really simple free-list style.

Comment (phadej) [reply src/Data/Arena.hs:71]: > That way the underlying allocator only has to deal with a single size of allocation, and so we can use a really simple free-list style.

There is no issue with supporting arbitrary allocation/alignment size though.

If the allocator reserves e.g. 1M, and then just keeps the pointer to the "rest/free" space, it's not difficult.

Recall, this allocator doesn't support individuals `free`s. Everything is released at once. And that what makes it simple (and hopefully fast).

Comment (dcoutts) [reply src/Data/Arena.hs:71]: > Should we change that first then?

With your flexible arena allocator, it would be possible to support the existing `indexSearches`, so we could introduce the changes in either order.

File: src/Database/LSMTree/Internal.hs [hunk 2]
```diff
@@ -512,9 +513,11 @@ lookups ks th fromEntry = withOpenTable th $ \thEnv -> do
     wb <- readMVar (tableWriteBuffer thEnv)
     let resolve = resolveMupsert (tableConfig th)
     cache <- readMVar (tableCache thEnv)
+    arenaManager <- newArenaManager -- TODO: use shared arena manager in handle (or session?)
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply src/Database/LSMTree/Internal.hs:540]: So here's my current thought (for use in future PRs):

Put the cache in the handle, but share the handle on `duplicate`. Only create a new independent cache in `new` and snapshot open.

This way independent tables will use independent caches and thus tables with different lifetimes will not affect the peak cache size in an extreme way. We'll not be able to be very aggressive with cache shrinking, if we have any shrinking at all.

**Review by dcoutts — APPROVED** [review]

File: src/Database/LSMTree/Internal.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (phadej) [reply src/Database/LSMTree/Internal.hs:540]: For the record: this sounds like we should not have independent tables in the single session to begin with. Then the cache will naturally be session-wide thing, instead of some intermediate thing between session and table-handle.
